be/src/exec/olap_scan_prepare.cpp

```diff
@@ -171,23 +179,192 @@ static bool get_predicate_value(ObjectPool* obj_pool, const SlotDescriptor& slot
+        RETURN_IF_ERROR(build_column_expr_predicates());
+    }
+
+    ASSIGN_OR_RETURN(auto normalized, _normalize_compound_predicates());
+    if (_is_root_builder) {
+        return normalized;
+    }
+    // Non-root builder return true only when all the child predicates are normalized.
+    return normalized && !SIMD::contain_zero(_normalized_exprs);
+}
+
+template <ExprContainer E, CompoundNodeType Type>
+StatusOr<bool> ChunkPredicateBuilder<E, Type>::_normalize_compound_predicates() {
+    const size_t num_preds = _exprs.size();
+    for (size_t i = 0; i < num_preds; i++) {
+        if (_normalized_exprs[i]) {
+            continue;
+        }
+
+        ASSIGN_OR_RETURN(const bool normalized, _normalize_compound_predicate(_exprs[i].root()));
+        if (!normalized && !_is_root_builder) {
+            return false;
+        }
+        _normalized_exprs[i] = normalized;
+    }
+
+    return true;
+}
+
+template <ExprContainer E, CompoundNodeType Type>
+StatusOr<bool> ChunkPredicateBuilder<E, Type>::_normalize_compound_predicate(const Expr* root_expr) {
+    auto process = [&]<CompoundNodeType ChildType>() -> StatusOr<bool> {
+        ChunkPredicateBuilder<RawExprContainer, ChildType> child_builder(
+                _opts, build_raw_expr_containers(root_expr->children()), false);
+        ASSIGN_OR_RETURN(const bool normalized, child_builder.parse_conjuncts());
+        if (normalized) {
+            _child_builders.emplace_back(std::move(child_builder));
+        }
+        return normalized;
+    };
+
+    if (TExprOpcode::COMPOUND_OR == root_expr->op()) {
+        if (!_opts.pred_tree_params.enable_or) {
+            return false;
+        }
+        return process.template operator()<CompoundNodeType::OR>();
+    }
+
+    if (TExprOpcode::COMPOUND_AND == root_expr->op()) {
+        return process.template operator()<CompoundNodeType::AND>();
+    }
+
+    return false;
+}
+
+template <ExprContainer E, CompoundNodeType Type>
+StatusOr<PredicateCompoundNode<Type>> ChunkPredicateBuilder<E, Type>::get_predicate_tree_root(
+        PredicateParser* parser, ColumnPredicatePtrs& col_preds_owner) {
+    auto compound_node = PredicateCompoundNode<Type>{};
+
+    const size_t start_i = col_preds_owner.size();
+    RETURN_IF_ERROR(_get_column_predicates(parser, col_preds_owner));
+    for (size_t i = start_i; i < col_preds_owner.size(); i++) {
+        compound_node.add_child(PredicateColumnNode{col_preds_owner[i].get()});
+    }
+
+    for (auto& child_builder : _child_builders) {
+        RETURN_IF_ERROR(std::visit(
+                [&](auto& c) {
+                    ASSIGN_OR_RETURN(auto child_node, c.get_predicate_tree_root(parser, col_preds_owner));
+                    compound_node.add_child(std::move(child_node));
+                    return Status::OK();
+                },
+                child_builder));
+    }
+
+    return compound_node;
+}
+
+template <bool Inverted>
+static bool is_not_in(const auto* pred) {
```

Comment:
inverted make the code hard to read, suggest to provide  two function: is_not_in, is_in?

Reply:
`is_not_in<Inverted>` is called by other methods with `<Inverted>`, they could use `is_not_in<Inverted>` without judging call `is_not_in` or `is_in`. 

If it is divided into two methods `is_not_in` and `is_in`, then they need to judge whether they should call `is_not_in` or `is_in`.

Maybe after renaming it to `is_not_in<Negative>`, it will be clearer?